### PR TITLE
fix: change the principal of reporting not working

### LINKED ISSUE #419: [Reporting] Dashboard not recreate in new user after upgrade pipeline
## Summary

After upgrading the pipeline, a new user parameter was passed to the reporting stack, but the dashboard was not recreated to a new user.
截屏2023-11-03 20 01 42

## Possible fixes

---

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by llmin**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

fix #419 

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [x] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/reporting/private/dashboard.ts [hunk 1]
```diff
@@ -86,6 +86,37 @@ export const dataSetActions = [
...
+
+export const CLICKSTREAM_RETENTION_VIEW_PLACEHOLDER = 'clickstream_retention_view';
+export const CLICKSTREAM_SESSION_VIEW_PLACEHOLDER = 'clickstream_session_view';
+export const CLICKSTREAM_USER_DIM_VIEW_PLACEHOLDER = 'clickstream_user_dim_view';
+export const CLICKSTREAM_EVENT_VIEW_PLACEHOLDER = 'clickstream_event_view';
+export const CLICKSTREAM_DEVICE_VIEW_PLACEHOLDER = 'clickstream_device_view';
+export const CLICKSTREAM_EVENT_PARAMETER_VIEW_PLACEHOLDER = 'clickstream_event_parameter_view';
+export const CLICKSTREAM_LIFECYCLE_DAILY_VIEW_PLACEHOLDER = 'clickstream_lifecycle_daily_view';
+export const CLICKSTREAM_LIFECYCLE_WEEKLY_VIEW_PLACEHOLDER = 'clickstream_lifecycle_weekly_view';
+export const CLICKSTREAM_USER_ATTR_VIEW_PLACEHOLDER = 'clickstream_user_attr_view';
```
Comment (zxkane) [inline src/reporting/private/dashboard.ts:118]: Why introduce the duplicated and unused constnat again?

https://github.com/awslabs/clickstream-analytics-on-aws/blob/c67d3c1956d89c39351d4f2cecc4e55ef06a9ad6/src/common/constant.ts#L345

File: src/reporting/private/dashboard.ts [hunk 2]
```diff
@@ -105,70 +136,58 @@ export async function waitForDataSetCreateCompleted(quickSight: QuickSight, acco
...
       }
-      logger.info('AnalysisCreate: sleep 1 second');
+
+      logger.info('AnalysisUpdate: sleep 1 second');
       await sleep(1000);
 
     } catch (err: any) {
-      logger.info('AnalysisCreate catch: sleep 1 second');
-      await sleep(1000);
+      throw err;
```
Comment (zxkane) [inline src/reporting/private/dashboard.ts:165]: It's unnecessary to catch the error while doing nothing.

File: src/reporting/private/dashboard.ts [hunk 3]
```diff
@@ -105,70 +136,58 @@ export async function waitForDataSetCreateCompleted(quickSight: QuickSight, acco
...
+        throw new Error('Dashboard update failed.');
       }
-      logger.info('DashboardCreate: sleep 1 second');
+      logger.info('DashboardUpdate: sleep 1 second');
       await sleep(1000);
 
     } catch (err: any) {
-      logger.info('DashboardCreate catch: sleep 1 second');
-      await sleep(1000);
+      throw err;
```
Comment (zxkane) [inline src/reporting/private/dashboard.ts:190]: It's unnecessary to catch the error while doing nothing.

File: src/reporting/lambda/custom-resource/quicksight/index.ts [hunk 4]
```diff
@@ -170,9 +172,16 @@ const _onUpdate = async (quickSight: QuickSight, awsAccountId: string, principal
     const dashboardDefProps: QuickSightDashboardDefProps = props.dashboardDefProps;
     logger.info('dashboardDefProps', JSON.stringify(dashboardDefProps));
 
+    let newPrincipalArn = undefined;
+    if (oldProps.quickSightUser !== props.quickSightUser) {
```
Comment (zxkane) [inline src/reporting/lambda/custom-resource/quicksight/index.ts:176]: Add test for covering this case where the principal is changed when updating the template

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/reporting/lambda/custom-resource/quicksight/index.ts (hunk 4 above)
Comment (llmin) [reply src/reporting/lambda/custom-resource/quicksight/index.ts:176]: already added.  <Update QuickSight dashboard - permission update>

File: src/reporting/private/dashboard.ts [hunk 5]
```diff
@@ -105,70 +126,60 @@ export async function waitForDataSetCreateCompleted(quickSight: QuickSight, acco
...
-      if ( template.Template !== undefined && template.Template?.TemplateId !== undefined ) {
-        return;
-      }
-      logger.info('TemplateCreate: sleep 1 second');
-      await sleep(1000);
-
-    } catch (err: any) {
-      logger.info('TemplateCreate catch: sleep 1 second');
-      await sleep(1000);
+      throw err;
```
Comment (zxkane) [inline src/reporting/private/dashboard.ts:129]: It's unnecessary to catch the error while doing nothing.

File: test/reporting/quicksight/lambda/quicksight-resource.test.ts [hunk 6]
```diff
@@ -558,6 +647,75 @@ describe('QuickSight Lambda function', () => {
 
   });
 
+  test('Update QuickSight dashboard - permission update', async () => {
```
Comment (zxkane) [inline test/reporting/quicksight/lambda/quicksight-resource.test.ts:650]: also assert the new principal is used in API calls

**Review by zxkane — CHANGES_REQUESTED** [review]

**Review by zxkane — APPROVED** [review]

**Pull request merged**